proc/compiler/compiler.go

```diff
@@ -117,14 +120,43 @@ func compileProc(custom Hook, node ast.Proc, pctx *proc.Context, parent proc.Int
 	case *ast.FuseProc:
 		return fuse.New(pctx, parent)
 
+	case *ast.JoinProc:
+		return nil, ErrJoinParents
+
 	default:
 		return nil, fmt.Errorf("unknown AST type: %v", v)
 
 	}
 }
 
+func compileJoin(custom Hook, v *ast.JoinProc, rest []ast.Proc, pctx *proc.Context, left, right proc.Interface) ([]proc.Interface, error) {
+	join, err := join.New(pctx, left, right, v)
+	if err != nil {
+		return nil, err
+	}
+	parents := []proc.Interface{join}
+	if len(rest) == 0 {
+		return parents, nil
+	}
+	return compileSequential(custom, rest, pctx, parents)
+}
+
+func enteringJoin(nodes []ast.Proc) bool {
+	var ok bool
+	if len(nodes) > 0 {
+		_, ok = nodes[0].(*ast.JoinProc)
+	}
+	return ok
+}
+
 func compileSequential(custom Hook, nodes []ast.Proc, pctx *proc.Context, parents []proc.Interface) ([]proc.Interface, error) {
 	node := nodes[0]
+	if join, ok := node.(*ast.JoinProc); ok {
+		if len(parents) != 2 {
+			return nil, ErrJoinParents
+		}
+		return compileJoin(custom, join, nodes[1:], pctx, parents[0], parents[1])
+	}
```

Comment:
This is surprising. Why not add a case for ast.JoinProc in Compile instead, as we have for ast.ParallelProc and ast.SequentialProc?